### PR TITLE
Cover smart contract update transaction costs.

### PR DESCRIPTION
**Pull request by eb-concordium**
## Purpose

Closes https://github.com/Concordium/concordium-wallet-proxy/issues/63.

## Changes

* transactionCost endpoint can now calculate the cost of smart contract update transactions.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -242,6 +243,11 @@ The following query parameters are supported
 - `transactionCommission`, whether the transaction fee commission of a baker pool is updated. Optionally, and only applies when `type` is either `updateBakerPool` or `configureBaker`.
 - `bakerRewardCommission`, whether the baker reward commission of a baker pool is updated. Optionally, and only applies when `type` is either `updateBakerPool` or `configureBaker`.
 - `finalizationRewardCommission`, whether the finalization reward of a baker pool is updated. Optionally, and only applies when `type` is either `updateBakerPool` or `configureBaker`.
+- `invoker`, only applies when `type` is `update` and is mandatory in this case. Specifies the sender account address of the transaction.
```
Comment (abizjak) [inline README.md:246]: Perhaps we should call this `sender` instead, to make it more understandable.

File: README.md [hunk 2]
```diff
@@ -242,6 +243,11 @@ The following query parameters are supported
...
+- `contractIndex`, only applies when `type` is `update` and is mandatory in this case. Specifies the smart contract index of the contract being updated.
+- `contractSubindex`, only applies when `type` is `update` and is mandatory in this case. Specifies the smart contract subindex of the contract being updated.
+- `receiveName`, only applies when `type` is `update` and is mandatory in this case. Specifies the smart contract receive name of the smart contract receive function being called.
+- `parameter`, only applies when `type` is `update` and is mandatory in this case. Specifies the smart contract parameter passed to the receive function.
```
Comment (abizjak) [inline README.md:250]: Can you specify the encoding of this (hex).

File: README.md [hunk 3]
```diff
@@ -242,6 +243,11 @@ The following query parameters are supported
 - `transactionCommission`, whether the transaction fee commission of a baker pool is updated. Optionally, and only applies when `type` is either `updateBakerPool` or `configureBaker`.
 - `bakerRewardCommission`, whether the baker reward commission of a baker pool is updated. Optionally, and only applies when `type` is either `updateBakerPool` or `configureBaker`.
 - `finalizationRewardCommission`, whether the finalization reward of a baker pool is updated. Optionally, and only applies when `type` is either `updateBakerPool` or `configureBaker`.
+- `invoker`, only applies when `type` is `update` and is mandatory in this case. Specifies the sender account address of the transaction.
+- `contractIndex`, only applies when `type` is `update` and is mandatory in this case. Specifies the smart contract index of the contract being updated.
+- `contractSubindex`, only applies when `type` is `update` and is mandatory in this case. Specifies the smart contract subindex of the contract being updated.
```
Comment (abizjak) [inline README.md:248]: Can you specify the type (integer)?

File: src/Proxy.hs [hunk 4]
```diff
@@ -507,6 +512,76 @@ getTransactionCostR = withExchangeRate $ \(rate, pv) -> do
...
+                      Just receiveName -> if Wasm.isValidReceiveName receiveName
+                        then return $ Wasm.ReceiveName receiveName
+                        else respond400Error (EMParseError "Invalid receive name.") RequestInvalid
+                wasmParameter <- lookupGetParam "parameter" >>= \case
+                      Nothing -> respond400Error (EMParseError "Missing `parameter` value.") RequestInvalid
+                      Just parameterText -> case BS16.decode . Text.encodeUtf8 $ parameterText of
+                        Left s -> respond400Error (EMParseError $ "Could not parse `parameter` value: " ++ s) RequestInvalid
+                        Right parameter -> return $ Wasm.Parameter $ BSS.toShort parameter
+
+                let nrg = Energy 30000000 -- 30 million
```
Comment (abizjak) [inline src/Proxy.hs:552]: I think we should be much more conservative here, and not allow more than 500ms to execute the estimate. So set this to 500_000

**Review by abizjak — COMMENTED** [review]
Looks good overall.

Could you also add (to the documentation) an example query (say on testnet) since the parameters are quite complicated.

File: README.md (hunk 2 above)
Comment (eb-concordium) [reply README.md]: Done.

File: README.md (hunk 3 above)
Comment (eb-concordium) [reply README.md]: Done.

File: src/Proxy.hs (hunk 4 above)
Comment (eb-concordium) [reply src/Proxy.hs]: Done.

**Review by abizjak — APPROVED** [review]
Looks good.

**Pull request merged**